platform/linux-generic/odp_ipsec.c

```diff
@@ -4,179 +4,1244 @@
+		status->error.alg = 1;
+		goto out;
+	}
+
+	if (ODP_IPSEC_MODE_TUNNEL == ipsec_sa->mode) {
+		/* We have a tunneled IPv4 packet, strip outer and IPsec
+		 * headers */
+		odp_packet_move_data(pkt, ip_hdr_len + hdr_len, 0,
+				     ip_offset);
+		if (odp_packet_trunc_head(&pkt, ip_hdr_len + hdr_len,
+					  NULL, NULL) < 0) {
+			status->error.alg = 1;
+			goto out;
+		}
+	} else {
+		odp_packet_move_data(pkt, hdr_len, 0,
+				     ip_offset + ip_hdr_len);
+		if (odp_packet_trunc_head(&pkt, hdr_len,
+					  NULL, NULL) < 0) {
+			status->error.alg = 1;
+			goto out;
+		}
+	}
+
+	/* Finalize the IPv4 header */
+	if (odp_packet_len(pkt) > sizeof(*ip)) {
+		ip = odp_packet_l3_ptr(pkt, NULL);
+
+		if (ODP_IPSEC_MODE_TRANSPORT == ipsec_sa->mode)
+			ipv4_adjust_len(ip, -(hdr_len + trl_len));
+
+		ip->ttl -= ipsec_sa->dec_ttl;
+		_odp_ipv4_csum_update(pkt);
+	}
+
+	if (!status->error.all) {
+		odp_packet_hdr_t *pkt_hdr = odp_packet_hdr(pkt);
+
+		packet_parse_reset(pkt_hdr);
+		/* FIXME: select parsing layer basing on config? */
+		packet_parse_l3_l4(pkt_hdr, ODP_PKTIO_PARSER_LAYER_ALL,
+				   ip_offset, _ODP_ETHTYPE_IPV4);
+	}
+out:
+	*pkt_out = pkt;
+
+	return ipsec_sa;
 }
 
-int odp_ipsec_sa_destroy(odp_ipsec_sa_t sa)
+/* Generate sequence number */
+static inline
+uint32_t ipsec_seq_no(ipsec_sa_t *ipsec_sa)
 {
-	(void)sa;
+	return odp_atomic_fetch_add_u32(&ipsec_sa->out.seq, 1);
+}
+
+/* Helper for calculating encode length using data length and block size */
+#define ESP_ENCODE_LEN(x, b) ((((x) + ((b) - 1)) / (b)) * (b))
+
+static ipsec_sa_t *ipsec_out_single(odp_packet_t pkt,
+				    odp_ipsec_sa_t sa,
+				    odp_packet_t *pkt_out,
+				    odp_ipsec_out_opt_t *opt ODP_UNUSED,
+				    odp_ipsec_op_status_t *status)
+{
+	ipsec_sa_t *ipsec_sa = NULL;
+	uint32_t ip_offset = odp_packet_l3_offset(pkt);
+	_odp_ipv4hdr_t *ip = odp_packet_l3_ptr(pkt, NULL);
+	uint16_t ip_hdr_len = ipv4_hdr_len(ip);
+	odp_crypto_packet_op_param_t param;
+	unsigned stats_length;
+	int rc;
+	uint16_t ipsec_offset;   /**< Offset of IPsec header from
+				      buffer start */
+	uint8_t	iv[MAX_IV_LEN];  /**< ESP IV storage */
+	ipsec_aad_t aad;         /**< AAD, note ESN is not fully supported */
+	unsigned hdr_len;        /**< Length of IPsec headers */
+	unsigned trl_len;        /**< Length of IPsec trailers */
+	uint8_t  ip_tos;         /**< Saved IP TOS value */
+	uint8_t  ip_ttl;         /**< Saved IP TTL value */
+	uint16_t ip_frag_offset; /**< Saved IP flags value */
+	odp_crypto_packet_result_t crypto; /**< Crypto operation result */
+
+	ODP_ASSERT(ODP_PACKET_OFFSET_INVALID != ip_offset);
+	ODP_ASSERT(NULL != ip);
+
+	ip_tos = 0;
+	ip_ttl = 0;
+	ip_frag_offset = 0;
+
+	ipsec_sa = _odp_ipsec_sa_use(sa);
+	ODP_ASSERT(NULL != ipsec_sa);
+
+	/* Initialize parameters block */
+	memset(&param, 0, sizeof(param));
+
+	if (ODP_IPSEC_MODE_TRANSPORT == ipsec_sa->mode &&
+	    _ODP_IPV4HDR_IS_FRAGMENT(ip->frag_offset)) {
```

Comment:
_ODP_IPV4HDR_IS_FRAGMENT() expects the parameter word in host byte order. The conversion is missing here.